server/src/main/java/org/opensearch/index/store/remote/utils/TransferManager.java

```diff
@@ -84,17 +88,47 @@ private static FileCachedIndexInput createIndexInput(FileCache fileCache, BlobCo
         return AccessController.doPrivileged((PrivilegedAction<FileCachedIndexInput>) () -> {
             try {
                 if (Files.exists(request.getFilePath()) == false) {
-                    try (
+                    List<InputStream> inputStreamList = new ArrayList<>();
+                    long partSize = request.getFileInfo().partSize().getBytes();
+                    long blockStart = request.getPosition();
```

Comment:
Sorry. I did not realize we will have to move the entire block calculation logic down into this class.
The transfer manager should not be concerned with this calculation and should act like a utility to fetch the block instead.

This might require some rework on `FileCachedIndexInput`, but it will be useful even for remote index cases that we are building up to.

Reply:
A quick thought here (will need refinements) -  

Within `OnDemandSnapshotBlockIndexInput`, we can handle the part calculation logic as follows - 

```
protected IndexInput fetchBlock(int blockId) throws IOException {
        final String blockFileName = fileName + "." + blockId;

        final long blockStart = getBlockStart(blockId);
        final long blockEnd = blockStart + getActualBlockSize(blockId);

        // If the snapshot file is chunked, we must account for this by
        // choosing the appropriate file part and updating the position
        // accordingly.
        final int startPartNumber = (int) (blockStart / partSize);
        final long partStart = startPartNumber * partSize;

        // Detect here if there are two chunks to be visited.
        // Add `append=true` to the blob fetch request
        // Change the API to only download blobs and open later.
        final int endPartNumber = (int) (blockEnd / partSize);

        final long position = blockStart - partStart;
        final long length = blockEnd - blockStart;

        BlobFetchRequest blobFetchRequest = BlobFetchRequest.builder()
            .position(position)
            .length(length)
            .blobName(fileInfo.partName(startPartNumber))
            .directory(directory)
            .fileName(blockFileName)
            .build();
        IndexInput indexInput = transferManager.fetchBlob(blobFetchRequest);

        if (endPartNumber != startPartNumber) {
           // recalculate position and length
            blobFetchRequest = BlobFetchRequest.builder()
                .position(position)
                .length(length)
                .blobName(fileInfo.partName(endPartNumber))
                .directory(directory)
                .fileName(blockFileName)
                .appendToFile(true)
                .build();
           indexInput = transferManager.fetchBlob(blobFetchRequest);
        }

        return indexInput;
    }
```

and within TransferManager - 

```
private static FileCachedIndexInput createIndexInput(FileCache fileCache, BlobContainer blobContainer, BlobFetchRequest request) {
        // We need to do a privileged action here in order to fetch from remote
        // and write to the local file cache in case this is invoked as a side
        // effect of a plugin (such as a scripted search) that doesn't have the
        // necessary permissions.
        return AccessController.doPrivileged((PrivilegedAction<FileCachedIndexInput>) () -> {
            try {
                if (Files.exists(request.getFilePath()) == false || request.appendToFile() == true) {
                    try (
                        InputStream snapshotFileInputStream = blobContainer.readBlob(
                            request.getBlobName(),
                            request.getPosition(),
                            request.getLength()
                        );
                        OutputStream fileOutputStream = Files.newOutputStream(request.getFilePath(), request.appendToFile() ?APPEND : null);
                        OutputStream localFileOutputStream = new BufferedOutputStream(fileOutputStream)
                    ) {
                        snapshotFileInputStream.transferTo(localFileOutputStream);
                    }
                }
                final IndexInput luceneIndexInput = request.getDirectory().openInput(request.getFileName(), IOContext.READ);
                return new FileCachedIndexInput(fileCache, request.getFilePath(), luceneIndexInput);
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
        });
    }
```

Reply:
I tried this, but problem of looping every chunk from SnapshotBlockIndexInput is in TransferManger [here](https://github.com/opensearch-project/OpenSearch/blob/03e415e8956dbc63b014f6b49c10650d353c90ca/server/src/main/java/org/opensearch/index/store/remote/utils/TransferManager.java#L58C1-L66C12) we check if file cache already exists in location if it is  , then we don't create new filecache instead we use existing cache. So, in our case we are only fetching first chunk, and for all of next chunks instead of fetching new chunk, we are returning same old chunk

Reply:
Using List to collect all blobfetchrequests at once in a loop and sending the `List<BlobFetchRequest>` to TransferManager. Updated the logic in new commit to fix above issue